### ISSUE TITLE
Why Rstudio delete your R project without asking you?

### ISSUE THREAD

**Issue by LinaQiu** [STAT545-UBC/Discussion#292]
Hi @jennybc and TAs,

I just went through a very horrible experience. I finished my homework09 this morning, sorry for missing the deadline, I was super busy last week... But when I was about to submit all my homework09 to github, it just suddenly disappeared... I have no idea what happened! And I have no backup for this, cause I did not have internet last night, thus I was not able to push my files to github step by step... I was so regretful now... But anyone has any idea about where to find my homework 09 back, or anyone happened to have such a awful experience before? I have tried `Trash`, homeworkk09 was not there... And I just realized that I did not have a time machine before... I am using a MACBook now, version 10.10.2. I'd appreciate it very much if anyone can give any suggestion!!!! Thank you so much!

Lina

**Comment by jennybc** [comment]
I have had a lot of RStudio crashes lately, due to my unplanned migration to Mac OS 10.11.1 El Capitan. However, never ever have I lost any files or a Project. In general, I have not even lost unsaved changes. (This is new to me -- previously I had almost never had RStudio crash on me.)

I am not denying this happened to you! But suggest you look really hard for these files. I wonder if you are confused about where the Project lives? Are you searching your hard drive systematically, e.g. through the Finder, vs. just browsing around where you expect to find things?

**Comment by LinaQiu** [comment]
I know exactly where I put all these files (same file for my STAT547 homework), cause when I create this project, I remember I change the file path to STAT547 file folder. Thus, it should be at that file. However, I do have used Finder to search my hard drive thoroughly, but there was nothing. I really have no idea what happened, and I really did nothing... it feels so bad.. Is there any other method I can try? Thanks!

Lina

**Comment by ksamuk** [comment]
Can you describe a bit more about the exact steps you took before you lost the project? It existed on your hard drive, it was open in R studio, and then what?

**Comment by jennybc** [comment]
@kevinushey do you have any tips here? I think there is nothing to be done?

**Comment by LinaQiu** [comment]
@ksamuk I opened my local Github to try to push it, but then I realized I want to check again in case there would be any typos, so I opened the whole project again in R studio, and checked it. After I checked it, I closed the whole project again, and came to Github, used `Sync` to get the lastest local Homework09 version, then IT JUST GONE! And I came back to my Homework09 file to check, there was nothing. When I tried to open Rstudio again, it told me Homework09 was not existed... That's all I have done, and all what happened to me. TT any tips? Or what I can do is to redo the whole homework again? So sad...

Lina

**Comment by ksamuk** [comment]
A few more questions:

1. When you say "local Github", do you mean the Gtihub desktop application? 
2. "used Sync to get the lastest local Homework09 version", do you mean you were hoping to sync your local repo with the remote repo hosted on github? 
3. What did the application say when you pressed sync? 
4. Does the repo still exist when you look at it in the Github app?
5. Are you perhaps on a different branch? Can you try switching branches?

You might need to get one of us to look at your computer/repo.  Unless something very weird has happened, I doubt that the Github application would (by default) delete local changes. So your project might still be around, or perhaps buried somewhere.

Also, if you feel comfortable with the command line, have a look at this Gist:

How To Recover a Deleted File From Your Git Repository
https://gist.github.com/paulbooker/1650111

**Comment by samhinshaw** [comment]
I see you're using a macbook, do you use time machine? If so, even if you haven't backed up recently, you may have a local copy saved: https://support.apple.com/en-ca/HT204015

**Comment by LinaQiu** [comment]
Hi @ksamuk ,

Thank you so much for your reply. For your questions,
1. yeah, I meant the Github desktop application.
2. "used Sync to get the latest local Homework09 version", I only meant get changes I made on my local github, since I haven't update my work onto Github, and I haven't made any changes on the remote github either.
3. I opened the github before I first opened my whole project in Rstudio again, and at the first time I checked github, I did see my Homework09 files there. However, after I opened my project in Rstudio  and modified some of my file, then closed the project, when I came to github and pressed sync, the application said nothing, but my Homework09 files gone away. 
4. I do not quite understand this question. But I think the answer would be no, since I haven't commit anything about my Homework09, there is no repository for HW09 on my remote githup yet.
5. No, I am certain sure I am at the same branch, cause I only have one "Master" branch. TAT

Yes, if any of you can help me with this, I would be happy to bring my laptop to your office, so you can check it for me. That would be great! Since if I cannot find it, I probably need to redo it again, that would be very awful, since final time is so busy. TAT

Thanks!
Lina

**Comment by ksamuk** [comment]
OK, sounds someone needs to look at your laptop. Can you bring it by office hours on Tuesday?

**Comment by LinaQiu** [comment]
Hi @STAT545-UBC/samuel_hinshaw ,

I have already checked the time machine, but unfortunately, I did not set up my time machine before. Cause this laptop is from my laptop, I have not thought about to set a time machine until this happened today...TAT Definitely, I should do better at data backup. TAT SO SAD.

**Comment by LinaQiu** [comment]
Hi @ksamuk ,

OK, I would bring my laptop during the office hours on Tuesday, but will also keep trying to solve this annoying issue these days too.

Thanks,
Lina

**Comment by jennybc** [comment]
@kevinushey suggests this, when a TA can take a look:

> If the `.git` folder exists, they can always try searching through `git reflog` and checking there.

**Comment by jennybc** [comment]
@LinaQiu 

It is no consolation but I lost the entire lesson on building an R package when my computer was stolen. Writing it the second was much faster.

**Comment by LinaQiu** [comment]
Hi @jennybc and @ksamuk ,

Yes, it is much faster to write it second time. I already did that. And seems like I found the issue now. Last time, when I was trying to commit my Homework09 to remote github, it was a whole file, and github may not be able to push an unzipped file? So it just disappeared directly, with a empty folder left. Cause this time when I tried to push it, it happened again. But this time, I do backup my whole file, so when it disappeared again, I realized the reason. Thus I create a new file folder in my local github, then copied&pasted all my homework09 files there. And now, it can be pushed successfully.

But it was so unreasonable, even if github cannot push a whole unzipped file folder, it should not delete it directly. It is so annoying. Anyway, thank you so much for your help.

Besides @ksamuk , seems like I do not need to bother you during next Tuesday's office hour. Anyway, thanks indeed for all your help! Thank you!

Best,
Lina

**Comment by daattali** [comment]
@LinaQiu that sounds very strange, I'm sure there's some logical explanation to what's happening.  If you're on campus during office hours, I'd be curious to see what's going on

**Comment by LinaQiu** [comment]
@daattali Yes sure, I also want to know the real reason. It is so unreasonable to delete your whole file only because github cannot upload a whole folder. I'd like to drop by your office during the office hour next Tuesday to see what's really happening here. 

Best,
Lina

**Comment by daattali** [comment]
Cool, see you there

**Comment by jennybc** [comment]
@daattali Did you learn anything we should record here?

**Comment by daattali** [comment]
You scared me! I just checked my calendar, and it is indeed not yet Tuesday afternoon. Will update after office hours :smile:

**Comment by daattali** [comment]
What we know: bad things happen when you have a git repo inside another git repo.  @LinaQiu created her homework as a new git repo INSIDE of her main course repo, then used GitHub Desktop (Mac) to "sync" (whatever that means, I'm unfamiliar with GitHub Desktop terminology), and then it deleted her new nested repo.

When we created homework folders NOT as github repos, it worked fine.

Maybe in the future it should be made more clear that it's a big no-no to make a git repo inside another repo. And I personally would suggest not recommending GitHub Desktop, and just using RStudio or SourceTree instead, because GitHub Desktop doesn't use the standard commit/push/pull, which makes it harder to know what's going on.

@LinaQiu feel free to add any info, or close this issue if you think everything is fine now

**Comment by kevinushey** [comment]
I believe a 'sync' does the equivalent of `git pull && git push`, so if the contents of a nested repository were not committed to the parent repository, it would be lost.

Indeed, nesting git repositories will only cause you grief, unfortunately...

**Comment by daattali** [comment]
Thanks for weighing in Kevin. Yeah, submodules is not something we talked about and would prefer not to open that can of worms... :)

**Comment by LinaQiu** [comment]
Yeah, I must keep in mind to never include a repo inside anothe repo! It's really painful. Lol Anyway, thanks a lot for all of your help! @jennybc @daattali @ksamuk @kevinushey Thank you!!!

Best,
Lina

**Comment by jennybc** [comment]
I wrote this once and then ... forgot to post? Trying again.

I've updated our page on Git clients to be even less positive about GitHub Desktop, since it is partially implicated here.

For the record, I do already have strong advice about NOT creating a Git repo within a Git repo in the course material already:

http://stat545-ubc.github.io/packages04_foofactors-package-01.html#create-the-package

Maybe I'll also add something to troubleshooting.

**Comment by LinaQiu** [comment]
@jennybc  My bad. I clicked the "Create a Git Repo" last time for homework08, `creating your own package`, and then it automatically default to be clicked every time i create a new R project. And yes, I learned from @daattali yesterday, we should only create a new file for our assignment every time, no need to create a new R project. So that's probably the reason, I did not realize your remind `not put a repo inside another repo` in homework08 when I was doing homework09. TAT Anyway, will keep it in mind from now on! Thanks Jenny!
